fastar/core.py

```diff
@@ -149,27 +145,35 @@ def _toposort(self, box):
         (self, to_global_coords(b)) for b in static_box_finder(
             (local_child_counts == 0) & (getbox(self.state, box) != KNOWN), 1)]
     while childless_boxes:
-      arr, box = childless_boxes.pop()
-      sorted_boxes.append((arr, box))
-      invals, _, primitive, params, _ = arr.eqn
-      instarts, counts = backward_rules[primitive](box, *invals, **params)
-      for ival, istart, count in zip(invals, instarts, counts):
-        if isinstance(ival, LazyArray) and istart is not None:
-          ibox = istart, count.shape
-          to_iglobal_coords = lambda b: (np.add(istart, b[0]), b[1])
-          ival.child_counts.subtract(
-              ibox, count * (getbox(ival.state, ibox) != KNOWN))
-          ilocal_child_counts = ival.child_counts.get(ibox)
-          childless_boxes.extend(
-              [(ival, to_iglobal_coords(b))
-               for b in static_box_finder(
-                   (ilocal_child_counts == 0) &
-                   (getbox(ival.state, ibox) != KNOWN), 1)])
+      self.process_childless_box(childless_boxes, sorted_boxes)
     return sorted_boxes[::-1]
 
+  def process_childless_box(self, childless_boxes, sorted_boxes):
+    arr, box = childless_boxes.pop()
+    invals, _, primitive, params, _ = arr.eqn
+    instarts, counts, outslice_from_inslices = backward_rules[primitive](
+      box, *(v.shape for v in invals), **params)
+    def outslice_from_invals(invals):
+      inslices = (None if instart is None else
+                  lax.dynamic_slice(inval, instart, count.shape)
+                  for inval, instart, count in zip(invals, instarts, counts))
+      return outslice_from_inslices(*inslices)
+    sorted_boxes.append((arr, box, outslice_from_invals))
```

Comment:
I'm wondering if it might be slightly simpler if `outslice_from_invals` also updated `arr`'s cache. I.e. if the definition was something like
```python
def outslice_from_invals_thunk():
  invals_ = [val.cache if isinstance(val, LazyArray) else val for val in invals]
  inslices = (None if instart is None else
              lax.dynamic_slice(inval, instart, count.shape)
              for inval, instart, count in zip(invals_, instarts, counts))
  outslice = outslice_from_inslices(*inslices)
  arr.cache = lax.dynamic_update_slice(cache, outslice, box_start)
```
(note no arguments and no return value). Then the only info needed by `_getbox` would be the above function, so instead of
```python
sorted_boxes.append((arr, box, outslice_from_invals))
```
you could simply have
```python
sorted_boxes.append(outslice_from_invals_thunk)
```
This would also simplify `_getbox`, because all it would have to do is execute the thunks in order and nothing else. Does that make sense?

Reply:
If you agree, then it might also make sense to rename `sorted_boxes` to `sorted_thunks`, and you could rename `outslice_from_invals_thunk` to just `thunk`. I also agree with the renaming you proposed of `backward_rules` -> `dependency_rules`.

Reply:
Completely agree, this makes things much simpler! I picked slightly different names and indicated thinks via type annotations. Let me know what you think.

Reply:
@j-towns Also renamed backward to dependency rules.